### DISCUSSION TITLE
"Address already in use" when quickly run tinytuya script for two different devices

### DISCUSSION THREAD

**Discussion by castus** [jasonacox/tinytuya#320] [Q&A] [answered]
I have a problem when I want **to control two different devices at almost the same time**. I'm invoking a script as a python's `subprocess` like so `subprocess.run('python run_tt.py "arg1" "arg2" "arg3"', shell=True, text=True)`. So I run something like that:

```python
subprocess.run('python run_tt.py "arg1" "arg2" "device1_id"', shell=True, text=True)
subprocess.run('python run_tt.py "arg1" "arg2" "device2_id"', shell=True, text=True)
```

The script is simple:

```python
import sys
import tinytuya

light = sys.argv[1]
place = sys.argv[2]
tuya_id = sys.argv[3]

d = tinytuya.OutletDevice(dev_id=tuya_id)

if light == "false":
  d.set_value(20, False)
else:
  d.set_value(20, True)
```

The error:

```bash
Apr 06 04:47:44 rpi4 sudo[5266]:     client.bind(("", UDPPORT))
Apr 06 04:47:44 rpi4 sudo[5266]: OSError: [Errno 98] Address already in use
Apr 06 08:50:27 rpi4 sudo[14327]: Traceback (most recent call last):
Apr 06 08:50:27 rpi4 sudo[14327]:   File "/home/pi/workspace/pico-ble/run_tt.py", line 8, in <module>
Apr 06 08:50:27 rpi4 sudo[14327]:     d = tinytuya.OutletDevice(dev_id=tuya_id)
Apr 06 08:50:27 rpi4 sudo[14327]:   File "/home/pi/.local/lib/python3.9/site-packages/tinytuya/core.py", line 745, in __init__
Apr 06 08:50:27 rpi4 sudo[14327]:     bcast_data = find_device(dev_id)
Apr 06 08:50:27 rpi4 sudo[14327]:   File "/home/pi/.local/lib/python3.9/site-packages/tinytuya/core.py", line 527, in find_device
Apr 06 08:50:27 rpi4 sudo[14327]:     client.bind(("", UDPPORT))
Apr 06 08:50:27 rpi4 sudo[14327]: OSError: [Errno 98] Address already in use
```

I guess, it's something with using the same IP Address, but what's bothering me is that these two devices have different IPs associated, right? So what's the reason for this issue? Could you help?

**Comment by uzlonewolf** [comment, marked as answer]
The issue is you are not specifying the IP address of the device.  This causes both threads to try and listen for the device broadcast packets at the same time.  We can work around the "Address already in use" error by setting the SO_REUSEPORT socket option, however that will not help in your case as devices only broadcast once every 5 seconds thereby causing a random delay before your devices will react to the commands you're sending.  Instead, you will need to either include the IPs in your script, or do a scan from your main thread and pass the IPs to the subprocess.  To do a scan for just a few specific devices:

```python
import tinytuya.scanner

want_devices = ['eb...1', 'eb...2', ...]
found = tinytuya.scanner.devices( verbose=False, scantime=8, poll=False, byID=True, wantids=want_devices )
```

Simply run the above fairly early in the main thread to find the IPs for your devices.  Note that offline or not-broadcasting devices will not be found and will thus be missing from the `found` dict.

**Reply by castus** [reply]
Thanks for the reply. Unfortunately, when I add an IP address to the constructor, everything stops working (meaning a device is not responding, when I set DPs) :(

This won't work at all:

```python
import tinytuya

d = tinytuya.OutletDevice(dev_id="bf55405d047f481050e0a8",address="192.168.0.15")
d.set_value(20, False)
```

but this works as expected

```python
import tinytuya

d = tinytuya.OutletDevice(dev_id="bf55405d047f481050e0a8")
d.set_value(20, False)
```

The IP address is correct, it has been taken from the scanner method you mentioned earlier.
Did I miss something?

**Reply by uzlonewolf** [reply]
Oh, yes, I forgot you also must specify the protocol version as well.
```python
d = tinytuya.OutletDevice(dev_id="bf55405d047f481050e0a8",address="192.168.0.15", version=3.X) 
```

**Reply by castus** [reply]
I can confirm that's working. Thank you.

In the meanwhile, I searched for different solutions to my problem and I found one. I can spin up a simple Python server, where I can bind to these two devices using `set_socketPersistent(True)`. Then I can react to GET requests and control devices based on requests. I tested it and it worked fine. My only concern is how to deal with disconnection to devices. How much I can rely on the persisting sockets? When I should restart the server? In general how to ensure High Availability to that connection? Do you have some thoughts on that?

The server script is as follows:

```python
import sys
import tinytuya
from http.server import BaseHTTPRequestHandler, HTTPServer

d = tinytuya.OutletDevice(dev_id="bf55405d047f481050e0a8")
d1 = tinytuya.OutletDevice(dev_id="bf40c7d2bc4798aa95xyyc")
d.set_socketPersistent(True)
d1.set_socketPersistent(True)

hostName = "localhost"
serverPort = 8088

class MyServer(BaseHTTPRequestHandler):
    def do_GET(self):
        d.set_value(20, True)
        d1.set_value(20, True)
        self.send_response(200)
        self.send_header("Content-type", "text/html")
        self.end_headers()

if __name__ == "__main__":        
    webServer = HTTPServer((hostName, serverPort), MyServer)
    print("Server started http://%s:%s" % (hostName, serverPort))

    try:
        webServer.serve_forever()
    except KeyboardInterrupt:
        pass

    webServer.server_close()
    print("Server stopped.")
```

**Reply by uzlonewolf** [reply]
Persistent sockets will automatically get re-opened if the device closes them.  How often will you be sending them data?  Devices will close the connection if they have not received anything in 30 seconds (I usually send empty heartbeat packets every 8-10 seconds to keep the connection open).  In addition, I would expect that server to have problems once the socket buffer fills up since you are never reading the received data.

I'm doing something similar to what you're doing, but with 20-30 devices.  I ended up making the sockets non-blocking and dumping them all into select.select().  Although my server isn't ready for release yet, I link to it in https://github.com/jasonacox/tinytuya/discussions/312#discussioncomment-5467810 .

**Reply by castus** [reply]
Hi @uzlonewolf , quick question, should the `tinytuya.scanner.devices ` method with `wantids` array only return these devices?

For ex. `tinytuya.scanner.devices( verbose=False, scantime=8, poll=False, byID=True, wantids=want_devices )`

If yes, I noticed that it does not. I see all devices in response json.

**Comment by castus** [comment]
Thanks @uzlonewolf for your time in answering my questions.
